protocol/consensus/consensus.go

```diff
@@ -0,0 +1,114 @@
+package consensus
+
+import (
+	"github.com/relab/hotstuff"
+	"github.com/relab/hotstuff/core"
+	"github.com/relab/hotstuff/core/eventloop"
+	"github.com/relab/hotstuff/core/logging"
+	"github.com/relab/hotstuff/modules"
+	"github.com/relab/hotstuff/protocol/proposer"
+	"github.com/relab/hotstuff/protocol/voter"
+	"github.com/relab/hotstuff/protocol/votingmachine"
+	"github.com/relab/hotstuff/service/committer"
+)
+
+// Consensus provides a default implementation of the Consensus interface
+// for implementations of the ConsensusImpl interface.
+type Consensus struct {
+	eventLoop *eventloop.EventLoop
+	logger    logging.Logger
+	config    *core.RuntimeConfig
+
+	committer *committer.Committer
+
+	extHandler modules.ExtProposeHandler
+
+	voter         *voter.Voter
+	votingMachine *votingmachine.VotingMachine
+	proposer      *proposer.Proposer
+}
+
+// New returns a new Consensus instance based on the given Rules implementation.
+func New(
+	// core dependencies
+	eventLoop *eventloop.EventLoop,
+	logger logging.Logger,
+	config *core.RuntimeConfig,
+
+	// protocol dependencies
+	extHandler modules.ExtProposeHandler,
+	proposer *proposer.Proposer,
+	voter *voter.Voter,
+	votingMachine *votingmachine.VotingMachine,
+
+	// service dependencies
+	committer *committer.Committer,
+) *Consensus {
+	cs := &Consensus{
+		eventLoop: eventLoop,
+		logger:    logger,
+		config:    config,
+
+		extHandler:    extHandler,
+		proposer:      proposer,
+		voter:         voter,
+		votingMachine: votingMachine,
+
+		committer: committer,
+	}
+	cs.eventLoop.RegisterHandler(hotstuff.ProposeMsg{}, func(event any) {
+		p := event.(hotstuff.ProposeMsg)
+		cs.OnPropose(&p)
+	})
+	return cs
+}
+
+// OnPropose is called when receiving a proposal from a leader and returns true if the proposal was voted for.
+func (cs *Consensus) OnPropose(proposal *hotstuff.ProposeMsg) {
+	block := proposal.Block
+	// ensure that I can vote in this view based on the protocol's rule.
+	err := cs.voter.Verify(proposal)
```

Comment:
This verify contains different parts.
1. It verifies that the proposal is well formed (crypto and from correct leader)
2. It verifies that we can vote for it (vote rule and last voted)
It is a change from previous logic, if blocks that pass 1 but not 2 are not handled for committing and advancing highQC. If these blocks are not stored in the blockchain, it may be necessary to later fetch them. 
However, this should happen only in problematic edge cases so not sure performance (additional fetch) is an issue here.

Reply:
@meling can you read this?